### PR TITLE
fix(backend_filesystem): skip 00_default.param when restarting after final step

### LINKED ISSUE #1507: Do not present 00_default.param after reaching 53_everyday_use.param
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Description of the bug

It presents an invalid file in the parameter editor.

### Steps To Reproduce

When the user completes the vehicle configuration by reaching the step "53_everyday_use.param" the program closes and stores the string "53_everyday_use.param" in the `last_uploaded_filename.txt`

Now when the program is restarted it will take a long time to load because it presents the file `00_default.param` in the GUI.

### Expected behavior

It must not do that. It must start from the first file that is NOT `00_default.param`.

### Screenshots / Logs

No

### Software information

- Operating system: Windows
- arduPilot Methodic Configuration version: 2.11.1

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tmchow**
## Summary

After the user reaches the final step (`53_everyday_use.param`), `get_start_file` now notices that the next index is past the end of the sequence and needs to wrap. Previously it returned `start_file`, which when `tcal_available=True` resolves to `files[0]` -- the read-only `00_default.param` snapshot -- so the next launch reopened the default-parameters dump and spent a long time loading it. Only the `00_default.param` case is rewritten; the existing `tcal_available=False` branch (already uses `files[2]`) is preserved. Addresses #1507.

## Why

`parameter_editor_and_uploader` writes the current step to `last_uploaded_filename.txt` on every upload. The moment the user finishes the full sequence, the "last uploaded" filename is literally `53_everyday_use.param`, and on relaunch `get_start_file`:

- computes `last_uploaded_index = files.index("53_everyday_use.param")`
- increments to `last_uploaded_index + 1`, which is past `len(files)`
- falls into the wraparound branch and returns `start_file`

`start_file` was computed earlier:

```python
if tcal_available:
    start_file = files[0]     # 00_default.param, the read-only factory snapshot
else:
    start_file = files[2]     # first non-default, non-tcal editable step
```

So vehicles with TCAL available reopen `00_default.param` every time -- exactly the broken behaviour reported in #1507. That file is large and is not an editable step, just a reference dump from the flight controller, which explains the "long time to load" observed by the reporter.

## Changes

Single narrow edit in `get_start_file`:

```python
if start_file_index >= len(files):
    logging_warning(_("Last uploaded file is the last file in the list. Starting from the beginning."))
    if start_file == "00_default.param":
        for candidate in files:
            if candidate != "00_default.param":
                return candidate
    return start_file
```

- The existing `tcal_available=False` path is untouched (`start_file` is already `files[2]`), which mirrors this path's historical intent.
- Only `start_file == "00_default.param"` triggers the forward walk -- that is the TCAL-available path, and it is the only path the issue report covers.
- The edge case where `00_default.param` is the only file left (someone manually trimmed the configuration) falls back to `start_file` rather than raising.

## Testing

- New `test_get_start_file_after_final_file_skips_default` covers all three branches:
  - TCAL available + last file reached -> `01_tcal.param` (not `00_default.param`).
  - TCAL not available + last file reached -> `02_first_real.param` (unchanged `files[2]` behaviour preserved).
  - Degenerate `files = {"00_default.param"}` -> returns `00_default.param` rather than looping forever.
- `python3 -m pytest tests/test_backend_filesystem.py` -> 72 passed locally.
- Confirmed the existing `test_get_start_file` still passes after the change, so the non-wraparound paths are unchanged.

Fixes #1507

This contribution was developed with AI assistance (Claude Code).

[Compound Engineering](https://github.com/EveryInc/compound-engineering-plugin)

### CODE REVIEW

File: ardupilot_methodic_configurator/backend_filesystem.py [hunk 1]
```diff
@@ -797,8 +797,16 @@ def get_start_file(self, explicit_index: int, tcal_available: bool) -> str:
...
+            # Respect the existing tcal_available branching (which already picks
+            # files[2] to skip 00_default + 01_tcal in the not-available path),
+            # and only adjust if start_file itself is 00_default.param -- that
+            # is a read-only factory snapshot, not an editable step (see #1507).
             logging_warning(_("Last uploaded file is the last file in the list. Starting from the beginning."))
+            if start_file == "00_default.param":
+                for candidate in files:
+                    if candidate != "00_default.param":
+                        return candidate
             return start_file
```
Comment (amilcarlucas) [inline ardupilot_methodic_configurator/backend_filesystem.py:810]: Why not just ?

```suggestion
            for candidate in files:
                if candidate != "00_default.param":
                    return candidate
            return start_file
```

**Comment by amilcarlucas** [comment]
@tmchow welcome to the project, thanks for your contribution.

File: tests/test_backend_filesystem.py [hunk 2]
```diff
@@ -579,6 +579,47 @@ def test_get_start_file(self) -> None:
...
+        with patch.object(lfs, "_LocalFilesystem__read_last_uploaded_filename", return_value="53_everyday_use.param"):
+            result = lfs.get_start_file(-1, tcal_available=False)
+        assert result == "02_first_real.param"
+
+        # When 00_default.param is the only remaining entry (edge case), fall back
+        # to the previously computed start_file rather than crashing.
+        lfs.file_parameters = {"00_default.param": {}}
+        with patch.object(lfs, "_LocalFilesystem__read_last_uploaded_filename", return_value="00_default.param"):
+            result = lfs.get_start_file(-1, tcal_available=True)
+        assert result == "00_default.param"
```
Comment (amilcarlucas) [inline tests/test_backend_filesystem.py]: In this edge case it should throw an error instead of silently let the user edit the non-editable 00_default.param file

**Comment by amilcarlucas** [comment]
to fix some of the issues run this on your computer

ruff format
ruff check --fix

**Comment by tmchow** [comment]
Thanks for the review @amilcarlucas. Latest commit applies both suggestions:

- **Line 807-810**: kept the `start_file == "00_default.param"` gate because removing it causes a regression for `tcal_available=False` - that path relies on `start_file = files[2]` to skip the tcal step, and an unconditional forward-walk would land on `01_tcal.param`. Simplified the loop body via `next((c for c in files if c != "00_default.param"), "")` which also drops a return statement (keeps PLR0911 happy).
- **Degenerate case**: `get_start_file` now `raise ValueError(...)` instead of silently returning `00_default.param` when that is the only entry. Test updated to assert the raise.

All `test_backend_filesystem.py` tests pass (72 passed) and `ruff check` / `ruff format` are clean.

**Review by amilcarlucas — APPROVED** [review]

**Comment by amilcarlucas** [comment]
Nice, thanks. Looks like you, or compound engineering read our CONTRIBUITING.md file.

**Comment by amilcarlucas** [comment]
Thanks, squashed and merged. Welcome to the project
